soccer/ui/MainWindow.cpp

```diff
@@ -544,234 +536,290 @@ void MainWindow::updateViews() {
-                } else if (rx.hardware_version() == RJ2015) {
-                    batteryLevel =
-                        RJ2015BatteryProfile.getChargeLevel(rx.battery());
-                } else {
-                    cerr << "Unknown hardware revision "
-                         << rx.hardware_version()
-                         << ", unable to calculate battery %" << endl;
+                // display error text
+                statusWidget->setErrorText(errorList.join(", "));
+
+                // show the ball in the robot's mouth if it has one
+                bool hasBall = rx.has_ball_sense_status() &&
+                               rx.ball_sense_status() == Packet::HasBall;
+                statusWidget->setHasBall(hasBall);
+
+                // battery
+                // convert battery voltage to a percentage and show it with the
+                // battery indicator
+                float batteryLevel = 1;
+                if (rx.has_battery()) {
+                    if (rx.hardware_version() == RJ2008 ||
+                        rx.hardware_version() == RJ2011) {
+                        batteryLevel = static_cast<float>(
+                            RJ2008BatteryProfile.getChargeLevel(rx.battery()));
+                    } else if (rx.hardware_version() == RJ2015 ||
+                               rx.hardware_version() == RJ2018) {
+                        batteryLevel = static_cast<float>(
+                            RJ2015BatteryProfile.getChargeLevel(rx.battery()));
+                    } else if (rx.hardware_version() == Simulation) {
+                        batteryLevel = 1;
+                    } else {
+                        cerr << "Unknown hardware revision "
+                             << rx.hardware_version()
+                             << ", unable to calculate battery %" << endl;
+                    }
                 }
-            }
-            statusWidget->setBatteryLevel(batteryLevel);
+                statusWidget->setBatteryLevel(batteryLevel);
 
-            // if there is an error bad enough that we should get this robot
-            // off the field, alert the user through the UI that there is a
-            // "showstopper"
-            bool showstopper = !hasVision || !hasRadio || hasMotorFault ||
-                               kickerFault || ballSenseFault ||
-                               (batteryLevel < 0.25) || !fpgaWorking;
-            statusWidget->setShowstopper(showstopper);
+                // if there is an error bad enough that we should get this robot
+                // off the field, alert the user through the UI that there is a
+                // "showstopper"
+                bool showstopper = !hasVision || !hasRadio || hasMotorFault ||
+                                   kickerFault || ballSenseFault ||
+                                   (batteryLevel < 0.25) || !fpgaWorking;
+                statusWidget->setShowstopper(showstopper);
 
 #endif
```

Comment:
This entire section is a monstrosity. Can you like break this out into multiple functions so it's readable?

Reply:
I'd been hoping to not have to do much in the way of trying to clean this up, because it's all about to be rewritten (@chachmu).